### PR TITLE
Autreg 68 refactoring

### PR DESCRIPTION
**Pull request by Eiriknil**

### CODE REVIEW

File: dynamo/src/main/java/no/bibsys/db/TableDriver.java [hunk 1]
```diff
@@ -78,80 +73,77 @@ public boolean tableExists(final String tableName) {
...
     /**
      * Return number of items in table
      * 
      * @param tableName
      * @return number of items
      */
-    
+
     public long tableSize(final String tableName) {
-
```
Comment (axthosarouris) [inline dynamo/src/main/java/no/bibsys/db/TableDriver.java]: One scan does not cover the whole database. It covers max. 1MB of data. It can be used to check if the database is empty but not to estimate the size of the table.
https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_Scan.html

File: dynamo/src/main/java/no/bibsys/db/TableDriver.java [hunk 2]
```diff
@@ -78,80 +73,77 @@ public boolean tableExists(final String tableName) {
...
-     * Deletes a table and the validation schema associated with this table
-     */
-    public void deleteTable(final String tableName) throws InterruptedException {
-        long itemCount = dynamoDb.getTable(tableName).describe().getItemCount();
+    public boolean deleteTable(final String tableName) {
+        
+        if(!tableExists(tableName)) {
+            return false;
+        }
+        ScanRequest scanRequest = new ScanRequest(tableName).withSelect(Select.COUNT);
```
Comment (axthosarouris) [inline dynamo/src/main/java/no/bibsys/db/TableDriver.java]: pack the checking if the table is empty in a function that is called isEmpty(), so that it is clear what we check

File: api/src/main/java/no/bibsys/db/EntityManager.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        return itemManager.updateItem(registryName, entity);
+    }
+
+    public boolean entityExists(String registryName, String entity) {
+        return itemManager.itemExists(registryName, entity);
+    }
+    
+    private String addIdToJson(final String json, final String entityId) throws IOException {
+        ObjectMapper objectMapper = ObjectMapperHelper.getObjectMapper();
+        JsonNode tree = objectMapper.readTree(json);
```
Comment (axthosarouris) [inline api/src/main/java/no/bibsys/db/EntityManager.java]: What happens if the json has  already an id field? Maybe it would be better if we had a form 

```
{ "id": "<id>", "body": { ... the rest of the json tree}}
        
```

File: api/src/main/java/no/bibsys/db/RegistryManager.java [hunk 4]
```diff
@@ -15,115 +14,99 @@
 
 public class RegistryManager {
 
-    private final transient TableDriver tableDriver;
+    private final transient TableDriver tableManager;
+    private final transient ItemDriver itemManager;
```
Comment (axthosarouris) [inline api/src/main/java/no/bibsys/db/RegistryManager.java]: change the variable names to tableDriver and itemDriver

File: api/src/main/java/no/bibsys/db/RegistryManager.java [hunk 5]
```diff
@@ -15,115 +14,99 @@
...
 
-    public void emptyRegistry(String tableName) throws InterruptedException {
-        TableManager tableManager = new TableManager(tableDriver);
-        tableManager.emptyTable(tableName);
+    public boolean registryExists(String tableName) {
+        return tableManager.tableExists(tableName);
+    }
 
+    public void emptyRegistry(String tableName) {
+        tableManager.deleteTable(tableName);
```
Comment (axthosarouris) [inline api/src/main/java/no/bibsys/db/RegistryManager.java]: Should be implemented as a function in tableDriver

File: api/src/main/java/no/bibsys/db/RegistryManager.java [hunk 6]
```diff
@@ -15,115 +14,99 @@
...
-            throw new TableNotFoundException(String.format("Registry %s does not exist", tableName));
-        }
+    public void deleteRegistry(String tableName) {
+
+        itemManager.deleteItem(getValidationSchemaTable(), tableName);
+        tableManager.deleteTable(tableName);
     }
 
-    public String getSchemaAsJson(String registryName) throws JsonParseException, JsonMappingException, IOException, InterruptedException {
+    public Optional<String> getSchemaAsJson(String registryName) throws JsonParseException, JsonMappingException, IOException, InterruptedException {
```
Comment (axthosarouris) [inline api/src/main/java/no/bibsys/db/RegistryManager.java]: there are redundant exceptions

File: api/src/main/java/no/bibsys/db/RegistryManager.java [hunk 7]
```diff
@@ -15,115 +14,99 @@
...
-        
+        EntityRegistryTemplate registryTemplate = mapper.readValue(registrySchemaItem.get(), EntityRegistryTemplate.class);
+
         String schema = Optional.ofNullable(registryTemplate.getSchema()).orElse("");
-        return schema;
+        return Optional.ofNullable(schema);
     }
-    
+
     public void setSchemaJson(String registryName, String schemaAsJson) throws JsonParseException, JsonMappingException, IOException, InterruptedException {
```
Comment (axthosarouris) [inline api/src/main/java/no/bibsys/db/RegistryManager.java]: redundant exceptions

File: dynamo/src/test/java/no/bibsys/db/ItemDriverTest.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    @Override
+    @Before
+    public void init() {
+        super.init();
+        itemDriver = ItemDriver.create(new DynamoDB(localClient));
+        tableDriver = TableDriver.create(localClient, new DynamoDB(localClient));
+    }
+
+    @Test
+    public void addAndGetItem() throws IOException {
```
Comment (axthosarouris) [inline dynamo/src/test/java/no/bibsys/db/ItemDriverTest.java]: tests should explain the expected behaviour. E.g. shouldThrowExceptionWhenAddingExistingId. 
Test if the add, update,etc behave well in corner cases,
Examples (more things should be checked):
add an existing item,
update an non existing item
delete an non existing item

File: dynamo/src/test/java/no/bibsys/db/TableDriverTest.java [hunk 9]
```diff
@@ -1,48 +1,110 @@
...
+    public void createTable() {
+        TableDriver tableDriver = newTableDriver();
+        tableDriver.createTable(template.getId());
+        List<String> tables = tableDriver.listTables();
+        int numberOfTables = tables.size();
+
+        assertThat(numberOfTables, is(equalTo(1)));
+    }
+
+    public void tableDriverWillNotCreateAnExistingTable() {
```
Comment (axthosarouris) [inline dynamo/src/test/java/no/bibsys/db/TableDriverTest.java]: Add the @Test label

File: api/src/main/java/no/bibsys/db/RegistryManager.java (hunk 5 above)
Comment (Eiriknil) [reply api/src/main/java/no/bibsys/db/RegistryManager.java]: Registry is not a table, it's a table + item in the registry schema table

File: api/src/main/java/no/bibsys/db/EntityManager.java (hunk 3 above)
Comment (Eiriknil) [reply api/src/main/java/no/bibsys/db/EntityManager.java]: put replaces any existing value

File: dynamo/src/main/java/no/bibsys/db/structures/SimpleEntry.java [hunk 10]
```diff
@@ -37,5 +41,7 @@ public void setPreferredLabels(final List<LanguageString> labels) {
         this.preferredLabels = labels;
     }
 
-
+    public String asJson() throws JsonProcessingException {
+        return ObjectMapperHelper.getObjectMapper().writeValueAsString(this);
```
Comment (axthosarouris) [inline dynamo/src/main/java/no/bibsys/db/structures/SimpleEntry.java]: Maybe should be declared in the interface? Since it is declared both in IdOnlyEntry and SImpleEntry

Comment (axthosarouris) [reply dynamo/src/main/java/no/bibsys/db/structures/SimpleEntry.java]: Also, SImpleEntry is used only in tests. It was quite stupid of me to add them to the production code, maybe more appropriate to move this class  along with `LanguageString` to tests?

File: dynamo/src/test/java/no/bibsys/db/TableDriverTest.java [hunk 11]
```diff
@@ -1,48 +1,101 @@
...
+
+public class TableDriverTest extends LocalDynamoTest {
+
+    @Test
+    public void createTable_TableNotExisting_AddsTable() {
+        TableDriver tableDriver = newTableDriver();
+        tableDriver.createTable(template.getId());
+        List<String> tables = tableDriver.listTables();
+        int numberOfTables = tables.size();
+
```
Comment (axthosarouris) [inline dynamo/src/test/java/no/bibsys/db/TableDriverTest.java]: Should we check if the method returns true here?
